server/notebooks/tasks.py

```diff
@@ -0,0 +1,46 @@
+import pytz
+from datetime import datetime, timedelta
+from collections import defaultdict
+from spinach import Tasks
+
+from .models import NotebookRevision
+
+TIME_INTERVAL_IN_MINS = 1
+tasks = Tasks()
+
+
+@tasks.task(name="notebooks:execute_notebook_revisions_cleanup")
+def execute_notebook_revisions_cleanup(notebook_id, now_utc=None):
+    "Prue revision hisotry."
+
+    def group_by_time_interval(notebook_revisions, time_interval_in_mins=TIME_INTERVAL_IN_MINS):
+        "Split notebook revisions into groups by time_interval_in_mins."
+        revision_groups = defaultdict(list)
+        for revision in notebook_revisions:
+            key = int(revision.created.timestamp()) // (time_interval_in_mins * 60)
+            revision_groups[key].append(revision)
+        return [revision_groups[k] for k in sorted(revision_groups.keys())]
+
+    def clear_draft_flag(notebook_revisions, now_utc, threshold_in_mins=TIME_INTERVAL_IN_MINS):
+        "Set is_draft flag to False if the revision is older than the threshold_in_mins."
+        for revision in notebook_revisions:
+            if now_utc - revision.created > timedelta(minutes=threshold_in_mins):
+                revision.is_draft = False
+                revision.save()
+
+    def delete(notebook_revisions, now_utc, threshold_in_mins=TIME_INTERVAL_IN_MINS):
+        "Delete the revision if it is older than threshold_in_mins."
+        for revision in notebook_revisions:
+            if now_utc - revision.created > timedelta(minutes=threshold_in_mins):
+                revision.delete()
+
+    # ordered_draft_revisions is ordered by `created` time desc.
+    ordered_draft_revisions = NotebookRevision.objects.filter(
+        notebook_id=notebook_id,
+        is_draft=True)
+    revision_groups = group_by_time_interval(ordered_draft_revisions)
+    now_utc = now_utc or datetime.now(tz=pytz.utc)
+
+    for group in revision_groups:
+        clear_draft_flag(group[:1], now_utc)
+        delete(group[1:], now_utc)
```

Comment:
I think it would be faster to just clear the draft flags in one operation at the end, rather than group-by-group. In that case you could simply filter for (remaining) revisions whose age is greater than the interval, and delete them in one go.

Reply:
Sure